docs/proposals/20210323-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,556 @@
+- User should be able to create a Kubernetes cluster that uses external etcd topology in a single step.
+- User should only have to create the CAPI Cluster resources, along with the required etcd provider specific resouces. That should trigger creation of an etcd cluster, followed by creation of the target workload cluster that uses this etcd cluster.
+- There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
+- Define a contract for pluggable etcd providers and define steps that controlplane providers should take to use a managed external etcd cluster.
+- Support following etcd cluster management actions: scale up and scale down, etcd member replacement, etcd version upgrades and rollbacks.
+- The etcd providers will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+- Etcd cluster members will undergo upgrades using the rollingUpdate strategy.
+
+### Non-Goals
+- The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available.
+- API changes such as adding new fields within existing ControlPlane providers, like the KubeadmControlPlane provider. We will utilize the existing external etcd endpoints field from the KubeadmConfigSpec for KCP.
+
+## Proposal
+
+### User Stories
+
+- As an end user, I want to be able to create a Kubernetes cluster that uses external etcd topology using CAPI with a single step.
+- On creating the required CRs in CAPI along with the new etcd provider specific CRs, CAPI should provision an etcd cluster, followed by a workload cluster that uses this external etcd cluster. 
+- As an end user, I should be able to use the etcd provider CRs to specify etcd cluster size during creation and specify/modify etcd version. CAPI controllers should modify and manage the etcd clusters accordingly.
+
+### Implementation Details/Notes/Constraints
+
+These are some of the key differences between etcd cluster provisioning flow when using etcdadm CLI vs the new etcdadm based provider:
+- Etcdadm cluster provisioning using CLI commands works as follows:
+
+  - Run `etcdadm init` on any one of the nodes to create a single node etcd cluster. This generates the CA cert & key on that node, along with the server, peer and client certs for that node.
+  - The init command also gives as output `etcdadm join` command with the client URL of the first node.
+  - To add a new member to the cluster, copy the CA cert key pair from the first node to the right location (`etc/etcd/pki`) on the new node. Then run the `etcdadm join <client URL>` command.
+  
+- This flow can't be used within cluster-api as is, since it requires copying over certs from the first etcd node to others. Instead it will follow the design that Kubeadm uses by generating certs outside the etcd cluster.
```

Comment:
How does this copy work in CAPI?

Reply:
So for KubeadmControlPlane controller in CAPI, instead of copying over certs, the controller generates certs and adds them to the `write_files` section of a cloud-init script. Each Machine is initialized with the cloud-init script and certs get added. I had written this section to explain how we'll use the same approach for the etcdadm bootstrap provider. But it might be confusing to follow so I have rephrased this section.